app/controllers/api/maintenance_stats_controller.rb

```diff
@@ -26,27 +26,24 @@ def begin_watching_bulk
   def begin_watching_repositories
     lookup_names = params.require(:repositories).map { |p| p.permit(%i[host_type full_name]).to_h.symbolize_keys }
 
-    github_lookup_names, other_lookup_names = lookup_names.partition { |repo_host_and_name| repo_host_and_name[:host_type].downcase == "github" }
+    auth_token = nil
+    repos = lookup_names.map do |repo_host_and_name|
+      # the create_from_host below does a remote API call to refresh the repo and THEN looks for an
+      # existing repo; we look for existing here first to save some time.
+      existing_repo = Repository::PersistRepositoryFromUpstream
+        .find_repository_from_host_type_and_full_name(host_type: repo_host_and_name[:host_type],
+                                                      full_name: repo_host_and_name[:full_name])
 
-    # grab a working token now so we don't run into issues trying to find one in the loop for each
-    # repository name
-    auth_token = AuthToken.find_token(:v3) if github_lookup_names.present?
-
-    # do we really want to synchronously create and refresh these repos here?
-    # Not sure this is the right choice.
-
-    github_repos = github_lookup_names.map do |repo_host_and_name|
-      Repository.create_from_host(repo_host_and_name[:host_type], repo_host_and_name[:full_name], auth_token.token)
-    end
-
-    other_repos = other_lookup_names.map do |repo_host_and_name|
-      Repository.create_from_host(repo_host_and_name[:host_type], repo_host_and_name[:full_name], nil)
+      existing_repo || begin
+        is_github = repo_host_and_name[:host_type].downcase == "github"
+        auth_token ||= AuthToken.find_token(:v3) if is_github
```

Comment:
re:
> grab a working token now so we don't run into issues trying to find one in the loop for each repository name

We're doing this now, is the thought that this will not be as much of an issue due to looking up existing repos first? We could still run into the same scenario if we start watching a bunch of GH repositories we don't know about at one (though probably rare)

Reply:
I didn't intend to change anything here in terms of behavior (auth_token is grabbed the first time we need it but then kept thereafter). I can't say I exactly understand what that comment was saying though.

Reply:
ah, yeah I missed the `||=`